ow_plexil/src/plexil-adapter/FaultDependencies.cpp

```diff
@@ -0,0 +1,355 @@
+#include "FaultDependencies.h"
+
+// ROS
+#include <ros/ros.h>
+
+// C++
+#include <unordered_map>
+#include <iomanip>      // std::setw
+#include <unordered_set>
+#include <stack>
+#include <vector>
+#include <iostream>
+#include <memory>
+
+// PUGIXML
+#include "pugixml.hpp"
+
+
+FaultDependencies::FaultDependencies(std::string file_name, bool verbose_flag)
+{
+  // If verbose we debug print after every update
+  m_verbose_flag = verbose_flag;
+  const char* source = file_name.c_str();
+  // Parse the dependency xml file in plans dir
+  parseXML(source);
+  // Initial print to show setup
+  DebugPrint();
+}
+
+
+bool FaultDependencies::checkIsOperable(const std::string &name){
+  // return true if subsystem/procedure is operable, false if inoperable
+  if (m_subsystems.find(name) != m_subsystems.end()){
+    return (m_subsystems[name].inoperable == 0);
+  }
+  else if (m_procedures.find(name) != m_procedures.end()){
+    return (m_procedures[name].inoperable == 0);
+  }
+  else{
+    ROS_WARN("IsOperable LOOKUP FAILED, %s DOES NOT EXIST.", name.c_str());
+    return true;
+  }
+}
+
+bool FaultDependencies::checkIsFaulty(const std::string &name){
+  // checks if subsystem is faulty
+  if (m_subsystems.find(name) != m_subsystems.end()){
+    return m_subsystems[name].faulty;
+  }
+  else{
+    ROS_WARN("IsFaulty LOOKUP FAILED, %s DOES NOT EXIST.", name.c_str());
+    return false;
+  }
+}
+
+
+std::vector<std::string> FaultDependencies::getActiveFaults(const std::string &name){
+  //LIMITING TO MAX OF 10 FAULTS RETURNED DUE TO PLEXIL RESTRICTIONS
+  std::vector<std::string> active_faults;
+
+  // If System is specified we return all active faults across every subsystem
+  if (name == "System"){
+    for (auto i: m_faults){
+      if (i.second.faulty == 1 && active_faults.size() <= 10){
+        active_faults.push_back(i.first);
+      }
+    }
+  }
+  // If no subsystem by name exists
+  else if(m_subsystems.find(name) == m_subsystems.end()){
+    ROS_WARN("SUBSYSTEM: %s DOES NOT EXIST.", name.c_str());
+  }
+  // Get all active faults in specified subsystem
+  else{
+    for (auto i: m_subsystems[name].faults){
+      if (m_faults[i].faulty == 1 && active_faults.size() <= 10){
+        active_faults.push_back(i);
+      }
+    }
+  }
+  return active_faults;
+}
+
+void FaultDependencies::updateFault(const std::string &name, int status){
+  // If a fault doesnt exist
+  if (m_faults.find(name) == m_faults.end()){
+    return;
```

Comment:
Would a warning be good here?

Reply:
Originally I had a warning here, but because this updateFault gets called for execution errors as well (something that is not in the XML), the output became verbose and confusing. This is due to it is using the LanderInterface.tpp to call updateFault.

It also lets the user not track certain faults in the XML without error messages.